Display Lifecycle hooks on the DC browse page

jhadvig opened the pull request:
I've created 3 previews on how we could display lifecycle hooks on the DC browse page. All three previews have similar structure, they only differ by the section where the particular hook description is  displayed and the position of the action that the hook is doing, which is either right from the hook type, or under the hook type (`text-muted` class)

1. As part of the **Configuration** section
![screenshot-67](https://cloud.githubusercontent.com/assets/1668218/20799291/022f3ed6-b7e2-11e6-8202-5192d7b9c125.png)
![screenshot-1](https://cloud.githubusercontent.com/assets/1668218/20799300/064f2ca6-b7e2-11e6-8d47-7a6422193b05.png)

2. Separate section in the `Details` tab
![screenshot-68](https://cloud.githubusercontent.com/assets/1668218/20799335/1e38f676-b7e2-11e6-9643-afa51a9f872b.png)
![screenshot](https://cloud.githubusercontent.com/assets/1668218/20799342/228822c4-b7e2-11e6-825d-f86fdbea1f1b.png)
Here I was not sure where to put the section so I've put it under the `Triggers` section

3. A separate `Lifecycle Hooks` tab since as you can see on the previous screens the `Details` tab can get pretty messy whe it contains a lot of things.
![screenshot-69](https://cloud.githubusercontent.com/assets/1668218/20799433/7e237f98-b7e2-11e6-8177-4d327a28e7f2.png)

IMHO I like the last option with the new tab, since it makes things more clean&clear. But I also understand that the hooks are more or less part of a "Configuration". On the other hand I kinda like the idea to split the some sections of the `Details` tab into a standalone tabs... eg: Autoscaling or Triggers ?!?

@jwforres @spadgett please let me know what what do you think. Please dont review the code for now ..was just prototyping. Thanks :)

spadgett commented:
I like having the separate tab. Shorten it to just "Hooks"?

I'm not sure about using the pod template styling for hooks. We currently use it as a distinct visual style for pod templates. Maybe something closer to the styling we use for dc details.

Otherwise looks good!

jhadvig commented:
yeah regarding similar styling that we have for the pod template.. I've reused it since it's using the concept of separating the details per container and thats basically what the lifecycle hooks do.

jwforres commented:
Honestly I prefer option 2.  I think if we were going to split stuff into another tab, we should really take a look at everything we have on that Details tab and come up with a logical split.  Hooks as its own tab doesn't make sense to me.  I'd expect a split more like (not necessarily this order):

How am I deploying:
- Strategy and params
- Hooks
- Scaling / Autoscaling
- Triggers

What am I deploying:
- Pod template details
- Volumes

jhadvig commented:
@spadgett @jwforres  here is preview of the styling we have in dc details in a separate tab:
![screenshot-2](https://cloud.githubusercontent.com/assets/1668218/20802749/56fbccb2-b7ed-11e6-88df-d5819e7a4c83.png)
![screenshot-3](https://cloud.githubusercontent.com/assets/1668218/20802748/56f827a6-b7ed-11e6-92a3-dd21a3cfb64a.png)

Here is preview of the styling we have in dc details in the Details tab:
![screenshot-4](https://cloud.githubusercontent.com/assets/1668218/20803736/87da95c2-b7f0-11e6-88b2-2fab3979b4c9.png)
![screenshot-5](https://cloud.githubusercontent.com/assets/1668218/20803737/87dd841c-b7f0-11e6-88e4-ca48880fdbf2.png)

jhadvig commented:
I liked the the pod-template like styling more, but dont have any major objection against the styling in the last comment as well (exepct https://github.com/openshift/origin-web-console/pull/966#issuecomment-264213424)

jhadvig commented:
I've updated the PR based on the conversation with @spadgett ..
Here is the screen on the how the Hooks will be displayed:
![screenshot-6](https://cloud.githubusercontent.com/assets/1668218/20840800/bcd8816e-b8b1-11e6-9659-e93e0de73862.png)

PTAL

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -1012,6 +1012,14 @@ angular.module('openshiftConsole')
+      if (!strategyName) {
+        return null;
+      }
+      return strategyName.toLowerCase() + 'Params';
```
We shouldn't assume this will always work. I'd rather have a switch statement for all the strategies we know about.

spadgett commented on app/views/browse/deployment-config.html:
```diff
@@ -129,17 +129,24 @@
-                              <dt>Max Surge:</dt>
-                              <dd>{{deploymentConfig.spec.strategy.rollingParams.maxSurge}}</dd>
+                              <dd>{{deploymentConfig.spec.strategy[strategyParams].timeoutSeconds}} sec</dd>
+                              <div ng-if="deploymentConfig.spec.strategy.rollingParams">
```
While you're in the code, can you change this to...?

```html
<dt ng-if-start="deploymentConfig.spec.strategy.rollingParams">Update Period:</dt>
<dd>{{deploymentConfig.spec.strategy.rollingParams.updatePeriodSeconds}} sec</dd>
...
<dt>Max Surge:</dt>
<dd ng-if-end>{{deploymentConfig.spec.strategy.rollingParams.maxSurge}}</dd>
```

Then you don't have the unnecessary div in the middle of the `dl`.

spadgett commented on app/views/browse/deployment-config.html:
```diff
@@ -198,6 +205,27 @@
+                        <div class="col-lg-6" ng-if="deploymentConfig.spec.strategy.type !== 'Custom'">
+                          <h3>
+                            Hooks
+                            <a href="{{'lifecycle_hooks' | helpLink}}" style="display: inline-block;" target="_blank">
```
Let's not use inline styles. Does `class="learn-more-inline"` work here?

spadgett commented on app/views/browse/deployment-config.html:
```diff
@@ -198,6 +205,27 @@
+                          <h3>
+                            Hooks
+                            <a href="{{'lifecycle_hooks' | helpLink}}" style="display: inline-block;" target="_blank">
+                              <span class="learn-more-block">Learn more <i class="fa fa-external-link"> </i></span>
```
Should be using `learn-more-inline` not `learn-more-block`.

Can you try

```html
<span class="learn-more-inline">
  <a ng-href="{{'lifecycle_hooks' | helpLink}}" target="_blank">Learn more&nbsp;<i class="fa fa-external-link" aria-hidden="true"></i></a>
</span>
```

spadgett commented on app/scripts/directives/lifecycleHook.js:
```diff
@@ -130,4 +130,15 @@ angular.module("openshiftConsole")
+      scope: {
+        hookParams: "=",
+        namespace: "=",
+        type: "@"
```
I think this is better as

```js
scope: {
  deploymentConfig: '=',
  type: '@'  // "pre", "mid", or "post"
}
```

Should be able to determine strategy from the deployment config spec, correct? Then you don't need to pass namespace because it's in the deployment config metadata.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+      <div class="is-item-description">
+        <dt>Command:</dt>
+        <dd>
+          <code class="probe-command">
```
I'm glad we're reusing these styles, but we should generalize the class name if it's not only used for health checks.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+    <small ng-switch-when="Ignore">(Continue with deployment on failure)</small>
+    <small ng-switch-when="Abort">(Abort deployment on failure)</small>
+    <small ng-switch-when="Retry">(Continue with deployment on failure)</small>
+  </span>
```
I don't think we should style this as help text since it's actual data in the hook and is not always the same value. I'd say we call it out specifically as failure policy below the header.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+          <dt>Environment Variables:</dt>
+          <dd>
+            <div ng-repeat="env in hookParams.execNewPod.env">
+              {{env.name}}&#8201;&#8594;&#8201;{{env.value}}
```
I think `{{env.name}}={{env.value}}` is more natural for environment variables.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+  <div ng-if="hookParams.execNewPod">
+    <h5 class="container-name">Container {{hookParams.execNewPod.containerName}}</h5>
+    <dl class="dl-horizontal left">
+      <div class="is-item-description">
```
I don't see where this CSS class is defined.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+    <div ng-repeat="tagImage in hookParams.tagImages">
+    <h5 class="container-name">Container {{tagImage.containerName}}</h5>
+      <dl class="dl-horizontal left">
+        <div class="is-item-description">
```
I don't see where this CSS class is defined.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
-              <istag-select model="istagHook" allow-custom-tag="true" select-disabled="view.isDisabled"></istag-select>
+<h4>
+  {{type}} Hook
+  <small class="text-muted">{{hookParams.execNewPod ? "Runs a command in a new pod based on the specified container" : "Tags image on deployment success"}}</small>
```
Same comment as failure policy: I don't think we should style this as help text because it's a configuration value that changes. I'd rather describe the pre/mid/post hook type in help text, then show `execNewPod` and `failurePolicy` clearly as configuration below the header.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,56 @@
+        <div class="is-item-description">
+          <dt>Tag As:</dt>
+          <dd>
+            <a ng-href="{{tagImage.to.name | navigateResourceURL : 'ImageStreamTag' : namespace}}">
```
Do you need to check first for `tagImage.to.namespace`? Then fall back `namespace` if not defined. (If `tagImage.to.namespace && tagImage.to.namespace !== namespace`, it should not be a link.)

jhadvig commented:
@spadgett I've updated the PR based on your comments.

Here is the screen of the Hooks in the DC browse page:
![screenshot-8](https://cloud.githubusercontent.com/assets/1668218/20921892/77dd2964-bba6-11e6-9c04-bd2f233bda54.png)

Wasnt really sure how to properly name the `execNewPod` and `tagImages` actions, so I named it `Action` and put its description next to it.
Also added a tooltip to the `failurePolicy` to explain what the value does.

PTAL

spadgett commented on app/scripts/directives/lifecycleHook.js:
```diff
@@ -130,4 +130,17 @@ angular.module("openshiftConsole")
+      },
+      templateUrl: 'views/directives/lifecycle-hook.html',
+      controller: function($scope) {
+        $scope.strategyParams = $filter('deploymentStrategyNameToParams')($scope.deploymentConfig.spec.strategy.type);
```
You should watch the type in case it changes.

```js
$scope.$watch('deploymentConfig.spec.strategy.type', function(type) {
  $scope.strategyParams = $filter('deploymentStrategyNameToParams')(type);
});
```

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -1012,6 +1012,20 @@ angular.module('openshiftConsole')
       return 'PAUSED_PENDING_INPUT' === stage.status;
     };
   })
+  .filter('deploymentStrategyNameToParams', function () {
```
I'd rather the filter give me the strategy params instead of the property name.

```js
.filter('strategyParams', function() {
  return function(deploymentConfig) {
    var strategy = _.get(deploymentConfig, 'spec.strategy.type');
    switch (strategy) {
    case 'Recreate':
      return _.get(deploymentConfig, 'spec.strategy.recreateParams', {});
    // ...
    default:
      return null;  
    }
  };
});
```

Then in the directive `link` function, you can get the strategy params with

```js
$scope.$watch('deploymentConfig', function(deploymentConfig) {
  $scope.strategyParams = $filter('strategyParams')(deploymentConfig);
});
```

(Should have made that comment on first review, sorry.)

spadgett commented on app/styles/_components.less:
```diff
@@ -188,7 +188,7 @@
 }
 
-code.probe-command {
+code.command {
```
👍

spadgett commented on app/views/browse/deployment-config.html:
```diff
@@ -198,6 +198,27 @@
+                          <h3>
+                            Hooks
+                            <span class="learn-more-inline">
+                              <a ng-href="{{'lifecycle_hooks' | helpLink}}" target="_blank">Learn more&nbsp;<i class="fa fa-external-link" aria-hidden="true"></i></a>
```
Should be "Learn More" (uppercase M) now. Changed the doc links in #978.

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,59 @@
+<span ng-switch="type">
+  <small class="text-muted" ng-switch-when="pre">Pre hooks execute before the deployment begins.</small>
+  <small class="text-muted" ng-switch-when="mid">Mid hooks execute after the previous deployment is scaled down to zero and before the first pod of the new deployment is created.</small>
+  <small class="text-muted" ng-switch-when="post">Post hooks execute after the deployment strategy completes.</small>
```
Instead of `small`, let's use the `help-block` style. (Margin looked off in the screenshot, but I think `help-block` will fix that.)

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,59 @@
+</span>
+<dl class="dl-horizontal left">
+  <dt>Action:</dt>
+  <dd>{{deploymentConfig.spec.strategy[strategyParams][type].execNewPod ? "Runs a command in a new pod based on the specified container" : "Tags image on deployment success"}}</dd>
```
I'd probably shorten these to "Run a command" and "Tag the image"

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,59 @@
+    <div ng-repeat="tagImage in deploymentConfig.spec.strategy[strategyParams][type].tagImages">
+    <h5 class="container-name">Container {{tagImage.containerName}}</h5>
+      <dl class="dl-horizontal left">
+        <dt>Tag As:</dt>
```
If there is only ever one item in this `dl`, I'd just put it in a `div`.

```html
<div>
  Tag image as <a href=...>myproject/image:latest</a>
</div>
```

jhadvig commented:
@spadgett comments addressed

jhadvig commented:
Adding screen:

![screenshot-12](https://cloud.githubusercontent.com/assets/1668218/20932674/b8c10da4-bbd5-11e6-8be3-1204b1d868df.png)

spadgett commented on app/views/directives/lifecycle-hook.html:
```diff
@@ -1,133 +1,58 @@
+  <span ng-switch="type">
+    <small ng-switch-when="pre">&ndash; runs before the deployment begins.</small>
+    <small ng-switch-when="mid">&ndash; runs after the previous deployment is scaled down to zero and before the first pod of the new deployment is created.</small>
+    <small ng-switch-when="post">&ndash; runs after the deployment strategy completes.</small>
```
nit: remove the period from the end of these messages

spadgett commented:
Instead of truncate long text, does `class="truncate"` work? Since it's a single line, I prefer to use CSS. Then it will truncate exactly at the edge of the screen.

spadgett commented:
If we switch to `class="truncate"`, set a `title` attribute with the full value as well.

jhadvig commented:
@spadgett updated :)
added the `dist` as well
PTAL

spadgett commented:
[merge]
